core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -49,14 +49,211 @@
+half_t cast_to_half(unsigned int val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned long val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned long long val);
+
+class half_t {
+  __half val;
+
+public:
+  KOKKOS_FORCEINLINE_FUNCTION
+  half_t():val(__float2half(0.f)){}
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  operator __half() const {
+    return val;
+  }
+
+// Changing below to 1 produces the constructor overload error
+#if 0
+  KOKKOS_FORCEINLINE_FUNCTION
+  operator bool() const {
+    return __half2float(val) == 0.0F;
+  }
+#endif
+
+  KOKKOS_INLINE_FUNCTION
+  half_t(__half rhs):val(rhs) {}
+
+#if 1
+  // Cast rhs to half for assignment to lhs of type half_t
+  template <class T>
+  KOKKOS_INLINE_FUNCTION
+  half_t(T rhs):val(cast_to_half(rhs)) {}
+#endif
+
+  // Unary operators
+  KOKKOS_INLINE_FUNCTION
+  half_t operator+ () const {
+    half_t tmp = *this;
+    #ifdef __CUDA_ARCH__
+      printf("half_t unary operator+\n");
+      tmp.val = +tmp.val;
+    #else
+      printf("float unary operator+\n");
+      tmp.val = __float2half(+__half2float(tmp.val));
+    #endif
+    return tmp;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  half_t operator- () const {
+    half_t tmp = *this;
+    #ifdef __CUDA_ARCH__
+      printf("half_t unary operator-\n");
+      tmp.val = -tmp.val;
+    #else
+      printf("float unary operator-\n");
+      tmp.val = __float2half(-__half2float(tmp.val));
+    #endif
+    return tmp;
+  }
+
+  // Binary operators
+  KOKKOS_INLINE_FUNCTION
+  half_t operator= (__half rhs) {
+    val = rhs;
+    return *this;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  half_t operator+ (__half rhs) const {
+    half_t tmp = *this;
+    #ifdef __CUDA_ARCH__
+      printf("half_t operator+\n");
+      tmp.val += rhs;
+    #else
+      printf("float operator+\n");
+      tmp.val = __float2half(__half2float(tmp.val) + __half2float(rhs));
+    #endif
+    return tmp;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  half_t operator- (__half rhs) const {
+    half_t tmp = *this;
+    #ifdef __CUDA_ARCH__
+      printf("half_t operator-\n");
+      tmp.val -= rhs;
+    #else
+      printf("float operator-\n");
+      tmp.val = __float2half(__half2float(tmp.val) - __half2float(rhs));
+    #endif
+    return tmp;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  half_t operator* (__half rhs) const {
+    half_t tmp = *this;
+    #ifdef __CUDA_ARCH__
+      printf("half_t operator*\n");
+      tmp.val *= rhs;
+    #else
+      printf("float operator*\n");
+      tmp.val = __float2half(__half2float(tmp.val) * __half2float(rhs));
+    #endif
+    return tmp;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  half_t operator/ (__half rhs) const {
+    half_t tmp = *this;
+    #ifdef __CUDA_ARCH__
+      printf("half_t operator/\n");
+      tmp.val /= rhs;
+    #else
+      printf("float operator/\n");
+      tmp.val = __float2half(__half2float(tmp.val) / __half2float(rhs));
+    #endif
+    return tmp;
+  }
+
+  // Logical operators
+  KOKKOS_INLINE_FUNCTION
+  half_t operator! () const {
```

Comment:
Do you actually need that one?

Reply:
I think so, !a and !!a for `float a` result in 0x0 regardless of what value a is.